File: 1544-count-good-nodes-in-binary-tree/1544-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        return self.recCheck(root, float('-inf'))

    def recCheck(self,root,big) -> int: 
        if not root:
            return 0
        good = True
        if root.val < big:
            good = False
        else:
            big = root.val
        leftCount = self.recCheck(root.left,big)
        rightCount = self.recCheck(root.right,big)
        
        return leftCount + rightCount + good
        

File: 1544-CountGoodNodesInBinaryTree/1544-CountGoodNodesInBinaryTree.cpp
// Last updated: 7/24/2025, 8:26:45 AM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
private:
    int traversal(TreeNode* root, int maxVal) {
        if (!root)
            return 0;
        if (root->val >= maxVal)
            return 1 + traversal(root->left, root->val) +
                   traversal(root->right, root->val);
        else
            return traversal(root->left, maxVal) +
                   traversal(root->right, maxVal);
    }

public:
    int goodNodes(TreeNode* root) {
        return traversal(root, root->val);
    }
};
auto init = []()
{ 
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 'c';
}();